type-generator/src/main/java/io/mesosphere/types/dtr/Main.java

```diff
@@ -0,0 +1,156 @@
+import java.io.PrintStream;
+import java.util.List;
+import java.util.Map;
+
+public class Main {
+
+    /**
+     * Entry point
+     * @param args
+     */
+    public static void main(String[] args) throws ParseException {
+        Options options = new Options();
+        CommandLineParser parser = new DefaultParser();
+
+        // Prepare options
+        Option o_help = Option.builder("h")
+                .longOpt("help")
+                .desc("Print this help message")
+                .argName("help")
+                .build();
+
+        Option o_project = Option.builder("p")
+                .longOpt("project")
+                .required()
+                .desc("The project name to target")
+                .hasArg()
+                .argName("project")
+                .build();
+
+        Option o_version = Option.builder("v")
+                .longOpt("version")
+                .required()
+                .desc("The project version to use")
+                .hasArg()
+                .argName("version")
+                .build();
+
+        Option o_repo = Option.builder("r")
+                .longOpt("repo")
+                .required()
+                .desc("The repository URI to use")
+                .hasArg()
+                .argName("repo")
+                .build();
+
+        // Compose options
+        options.addOption(o_help);
+        options.addOption(o_project);
+        options.addOption(o_version);
+        options.addOption(o_repo);
+
+        // Parse options
+        try {
+            CommandLine cmd = parser.parse(options, args);
+
+            // Handle and bail early on hep argument
+            if (cmd.hasOption("help")) {
+                HelpFormatter formatter = new HelpFormatter();
+                formatter.printHelp( "dtr", options );
+                return;
+            }
+
+            // Create a repository that is going to be our access
+            // channel to the type registry.
+            Repository repo = Repository.fromURI(cmd.getOptionValue("repo"));
+
+            // Access the specified project & version
+            RepositoryView project = repo.openProjectTypes(
+                cmd.getOptionValue("project"),
+                cmd.getOptionValue("version")
+            );
+
+            // Load project
+            RAMLParser raml = new RAMLParser();
+            GlobalScope globalScope = new GlobalScope();
+            raml.loadProjectTypes(project, globalScope);
+
+            // Initialize template
+            RepositoryView compilerView = repo.openProjectCompiler(
+                cmd.getOptionValue("project"),
+                cmd.getOptionValue("version")
+            );
+            CompilerEngine engine = new StringTemplateEngine(compilerView);
+            CompiledFragment frag = engine.getCompiledFragment(CompiledFragmentType.FILE_CONTENTS);
+
+            // Display type registry
+            for (TypeScope scope: globalScope.childScopesOfType(FileScope.class)) {
```

Comment:
Why do we have to iterator over the scopes? Couldn't the compiler do so?

Reply:
True, this needs a bit re-structuring. 👍  What happens there is the following:

* We create a `GlobalScope`
* For every RAML file we have we create a `FileScope`
* Technically, calling `GlobalScope.flatTypeMap` will suffice.

In a previous implementation the naming of the types was wrongly prefixed with the filename, thus I iterated over the child "FileScope"s. But I think we can remove this now.